maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelProcessor.java

```diff
@@ -95,8 +114,15 @@ public File locateExistingPom(File projectDirectory) {
 
     public Path locateExistingPom(Path projectDirectory) {
         // Note that the ModelProcessor#locatePom never returns null
-        File f = modelLocator.locateExistingPom(projectDirectory.toFile());
-        Path pom = f != null ? f.toPath() : null;
+        // while the ModelParser#locatePom needs to return an existing path!
+        Path pom = modelParsers.stream()
+                .map(m -> m.locate(projectDirectory).map(s -> s.getPath()).orElse(null))
+                .filter(Objects::nonNull)
+                .findFirst()
+                .orElseGet(() -> {
+                    File f = modelLocator.locateExistingPom(projectDirectory.toFile());
```

Comment:
probably don't need to convert to a File here, just use the Path

Reply:
`ModelLocator` is part of the Maven 3 public api and uses the `File` instead of `Path`.
The Maven 4 api uses `Path` and not `File`, but I don't think that's a good idea to start modifying the Maven 3 api in incompatible ways, or even to duplicate all the methods to leverage the nio API at this point.
So I'm not really sure what can be done here.

Reply:
@elharo I agree with @gnodet that checked exceptions should be used if calling code can do something about it (not necessarily "recover") and **should** do something, programming errors are usually nothing code can act uppon and will only result in try-catch-forget ...